File: go/wenmar/pagination.go
package wenmar

import (
	"context"
	"encoding/json"
	"io"
	"net/http"
	"net/url"
	"regexp"
	"strings"
)

var linkRelRE = regexp.MustCompile(`<([^>]+)>;\s*rel="([^"]+)"`)

func parseLinkHeader(header, rel string) string {
	if header == "" {
		return ""
	}
	matches := linkRelRE.FindAllStringSubmatch(header, -1)
	for _, m := range matches {
		if len(m) >= 3 && m[2] == rel {
			return m[1]
		}
	}
	return ""
}

type Paginator struct {
	nextURL string
	client  *Client
	// fetchNext is called with the raw next URL to fetch the next page.
	// It returns the response body and the Link header from the next response.
	fetchNext func(ctx context.Context, url string) (body []byte, linkHeader string, err error)
}

func (p *Paginator) HasNext() bool {
	return p.nextURL != ""
}

// NextPage fetches the next page by following the actual next URL from the
// Link header. The response is a generic decoded value (array or object).
func (p *Paginator) NextPage(ctx context.Context) (any, error) {
	if !p.HasNext() {
		return nil, nil
	}

	body, linkHeader, err := p.fetchNext(ctx, p.nextURL)
	if err != nil {
		return nil, err
	}

	// Advance to the next link from the response, if any.
	p.nextURL = parseLinkHeader(linkHeader, "next")

	// Decode the body as a generic value for the caller.
	var result any
	if err := json.Unmarshal(body, &result); err != nil {
		return nil, err
	}
	return result, nil
}

func newPaginatorFromResponse(resp *http.Response, client *Client) *Paginator {
	next := parseLinkHeader(resp.Header.Get("Link"), "next")
	return &Paginator{
		nextURL: next,
		client:  client,
		fetchNext: func(ctx context.Context, url string) ([]byte, string, error) {
			return client.fetchURL(ctx, url)
		},
	}
}

// sameOrigin reports whether rawURL has the same scheme, host, and port
// as baseURL. This prevents the SDK from sending credentials to a
// different host if the API returns a malicious Link header.
func sameOrigin(rawURL, baseURL string) bool {
	parsed, err := url.Parse(rawURL)
	if err != nil {
		return false
	}
	base, err := url.Parse(baseURL)
	if err != nil {
		return false
	}
	return strings.EqualFold(parsed.Scheme, base.Scheme) &&
		strings.EqualFold(parsed.Host, base.Host)
}

// fetchURL performs a raw GET against the given URL (which may include query
// params like ?page=2) and returns the response body and Link header.
// It validates the URL is same-origin as the client's BaseURL before
// attaching credentials, preventing token exfiltration via malicious
// Link headers.
func (c *Client) fetchURL(ctx context.Context, url string) ([]byte, string, error) {
	if !sameOrigin(url, c.BaseURL) {
		return nil, "", &APIError{
			Code:       "invalid_pagination",
			Message:    "pagination next URL is not same-origin as base URL",
			StatusCode: 0,
			Method:     "GET",
			Path:       url,
		}
	}

	req, err := http.NewRequestWithContext(ctx, "GET", url, nil)
	if err != nil {
		return nil, "", err
	}
	req.Header.Set("Authorization", "Bearer "+c.Token)
	req.Header.Set("Accept", "application/json")
	req.Header.Set("User-Agent", "wenmar-sdk-go")

	resp, err := c.http.Do(req)
	if err != nil {
		return nil, "", err
	}
	defer resp.Body.Close()

	if resp.StatusCode >= 400 {
		body, _ := io.ReadAll(resp.Body)
		return nil, "", ParseErrorBodyWithRequest(body, resp.StatusCode, "GET", url)
	}

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, "", err
	}

	return body, resp.Header.Get("Link"), nil
}

File: ruby/lib/wenmar/pagination.rb
# frozen_string_literal: true

require "uri"

module Wenmar
  class Paginator
    include Enumerable

    attr_reader :data, :links, :meta, :next_url, :client

    def initialize(client, page)
      @client = client
      @data = page.fetch("data", page.fetch(:data, []))
      @links = page.fetch("links", page.fetch(:links, {}))
      @meta = page.fetch("meta", page.fetch(:meta, {}))
      @page = page
      @next_url = @links["next"] || @links[:next]
    end

    def has_next?
      !@next_url.nil? && !@next_url.empty?
    end

    def next_page
      return nil unless has_next?

      fetch_next
    end

    def each
      return to_enum unless block_given?

      @data.each { |item| yield item }
      while has_next?
        data = fetch_next
        data.each { |item| yield item }
      end
    end

    def to_a(max = 1000)
      result = @data.dup
      result.concat(fetch_next) while has_next? && result.size < max
      result.first(max)
    end

    def self.parse_link_header(header, rel)
      return nil if header.nil? || header.empty?

      header.split(",").each do |part|
        match = part.match(/<([^>]+)>;\s*rel="#{rel}"/)
        return match[1] if match
      end
      nil
    end

    def self.from_response(response, client)
      link_header = response.headers["Link"]
      next_url = parse_link_header(link_header, "next")
      body = response.body
      parsed = body.is_a?(String) ? JSON.parse(body) : body
      data = parsed.is_a?(Array) ? parsed : []
      new(client, {"data" => data, "links" => {"next" => next_url}, "meta" => {}})
    end

    private

    def fetch_next
      unless same_origin?(@next_url, @client.base_url)
        raise Wenmar::Error.new(
          code: "invalid_pagination",
          message: "pagination next URL is not same-origin as base URL",
          field_errors: {},
          status: 0
        )
      end

      response = @client.send(:get_raw, @next_url)
      @next_url = self.class.parse_link_header(response.headers["Link"], "next")
      body = response.body
      parsed = body.is_a?(String) ? JSON.parse(body) : body
      parsed.is_a?(Array) ? parsed : []
    end

    def same_origin?(url, base_url)
      parsed = URI.parse(url)
      base = URI.parse(base_url)
      parsed.scheme == base.scheme && parsed.host == base.host && parsed.port == base.port
    rescue URI::InvalidURIError
      false
    end
  end
end
